### ISSUE TITLE
Some info on the newer matadata format

### ISSUE THREAD

**Issue by noembryo** [koreader/koreader#11063]
Well, the inevitable happened.. 😉 
The recent update broke the KOHighlights parsing of metadata files.
So, I have some questions for our friendly developers.

From the first look I took, I see that the main difference is that the header of the file is changed.
It seems to contain the path of the metadata file itself.
This might be very helpful in the future for moved metadata, nice addition.

The thing that most likely will give me trouble is that instead of the previous way of renaming the older file adding an `.old` extension, now you rename the *newer* file adding a `.new` extension. 😨 
That mean that I will now open the older file if I search for `.lua` files or two files for the same book if I add the `.new` extension to the allowed ones.
There is also a third option, to check for the existence of both in every folder, but this will add some overhead to directory scanning..
Is this something of transitionary measure or is the standard extension of the newer metadata.

As for the other fields, other than a new (?) `series_index`, I can't find any missing key (from the ones I use of course).
I must have missed a lot of conversations about this refactoring, but there was not a lot of free time.. 😞 
Are there any other changes that I missed or should I just fix these things and build an update ASAP?

**Comment by hius07** [comment]
Hello,
(1) There is no `.new` extension, there are no changes in this code.
(2) If `series` contains `#` and number, we separate it to `series_index`. No more changes in other fields.
(3) A user may edit the book metadata. It is saved in the file _custom_metadata.lua_ located in the usual sdr folder.
(4) You are aware of the new sdr location (hashdocsettings).

**Comment by hius07** [comment]
`.old` still works:
https://github.com/koreader/koreader/blob/b70f8666562470102ca91a872d940519f66c30ce/frontend/luasettings.lua#L252-L267

**Comment by noembryo** [comment]
>(1) There is no .new extension, there are no changes in this code.

Oupss! Well a user send me two zipped metadata files named .....lua and .....lua.new and I thought it was you guys... 😃 
Sorry about that, my bad! 😳 

>(2) If series contains # and number, we separate it to series_index. No more changes in other fields.

I haven't see it before. Does that mean that after the separation the `series` part is missing the number?
~~If so, I have to read it. If not, it's OK to have it with the series.~~
I saw that it remains in the series field, so that's OK, no new code here..

>(3) A user may edit the book metadata. It is saved in the file custom_metadata.lua located in the usual sdr folder.

And in this folder there are both `metadata.lua` *and* `custom_metadata.lua`?
If yes, which one are you updating.
Are they copies of each other or the `custom_metadata.lua` has only the edited fields?
It seems I'm not going to escape the extra checks. 😩 
>(4) You are aware of the new sdr location (hashdocsettings).

I've seen this name, but I'm not sure for what it refers to... 
Is it something like the older centralized `History` folder that was used in Android?

Thank you for the answers. 👍

**Comment by hius07** [comment]
> And in this folder there are both `metadata.lua` _and_ `custom_metadata.lua`?

Yes. The book metadata in `metadata.lua` is unchanged.
`custom_metadata.lua` contains original metadata and custom metadata.
```Lua
-- /storage/emulated/0/Books/AliceInWonderland.sdr/custom_metadata.lua
return {
    ["custom_props"] = {
        ["authors"] = "a",
        ["description"] = "d",
        ["keywords"] = "lw",
        ["language"] = "l",
        ["series"] = "s",
        ["series_index"] = 1,
        ["title"] = "t",
    },
    ["doc_props"] = {
        ["authors"] = "Lewis Carroll",
        ["keywords"] = "Fantasy fiction\
Children's stories\
Imaginary places -- Juvenile fiction\
Alice (Fictitious character from Carroll) -- Juvenile fiction",
        ["language"] = "en",
        ["pages"] = 107,
        ["title"] = "Alice's Adventures in Wonderland / Illustrated by Arthur Rackham. With a Proem by Austin Dobson",
    },
}

```

**Comment by hius07** [comment]
> I've seen this name, but I'm not sure for what it refers to...

https://github.com/koreader/koreader/pull/10945#issuecomment-1742962415

**Comment by noembryo** [comment]
>Yes. The book metadata in `metadata.lua` is unchanged.
`custom_metadata.lua` contains original metadata and custom metadata.

From what I can see in your sample `custom_metadata.lua` has a subset of `metadata.lua`.
Do they contain any `highlights` data.
If not, I don't have to parse them at all (since the book properties are both at `["doc_props"]` and `["stats"]` of the original `metadata.lua`.
BTW, should I keep reading these properties from the `["stats"]` section, or its going to be obsolete and I should switch to `["doc_props"]` section?

**Comment by NiLuJe** [comment]
> If not, I don't have to parse them at all (since the book properties are both at `["doc_props"]` and `["stats"]` of the original `metadata.lua`.

@hius07 might have fixed that duplication, too? (Or I'm misremembering things ;)).

**Comment by noembryo** [comment]
>#10945 (comment)

So, this is where I've seen it before, he he.. 😃 
So, I think I'll keep the same act, as long as I can.. 😇

**Comment by hius07** [comment]
`["doc_props"]` is enough.

Custom metadata doesn't contain highlights, only book metadata.
Once a user edited any field (say, corrected authors), the corrected value is displayed everywhere in KOReader (list of files, book information etc).
So a user may want to see it in KOHighlights as well.
Obviously, it's not of the highest importance.

**Comment by noembryo** [comment]
>`["doc_props"]` is enough.

Does that mean that the `["stats"]` section will be removed?
If not, I don't have to change the code..
>So a user may want to see it in KOHighlights as well.

Perhaps I should have both versions and offer an option?
I have to think about it. Maybe later..

**Comment by hius07** [comment]
> Does that mean that the `["stats"]` section will be removed?

No, it will not be removed, but it doesn't contain more metadata than `doc_props` does (except one case: when the title is empty `stats` has a filename without extension in that field).

> Perhaps I should have both versions and offer an option?

Within KOReader custom metadata prevails the original one everywhere.

**Comment by noembryo** [comment]
>No, it will not be removed, but it doesn't contain more metadata than `doc_props` does (except one case: when the title is empty stats has a filename without extension in that field).

Yes, I do use it for the extra info.
>Within KOReader custom metadata prevails the original one everywhere.

Hmm, I have to think about it. Maybe a future update.
I must roll out this one fast, because of the version changes and I will write something later..

**Comment by noembryo** [comment]
Fixed in [v1.7.3.0](https://github.com/noembryo/KoHighlights/releases/tag/v1.7.3.0)..
